### PR TITLE
[IMP] Flag business models

### PR DESCRIPTION
**Pull request by tde-banana-odoo**
Task: https://www.odoo.com/web#id=30151&view_type=form&model=project.task&action=333&active_id=248&menu_id=4720

### CODE REVIEW

File: odoo/models.py [hunk 1]
```diff
@@ -245,7 +246,6 @@ class BaseModel(object):
     _date_name = 'date'         # field to use for default calendar view
     _fold_name = 'fold'         # field to determine folded groups in kanban views
 
-    _needaction = False         # whether the model supports "need actions" (see mail)
```
Comment (rco-odoo) [inline odoo/models.py]: This has nothing todo with this task.
Please move this cleanup in a separate PR.

File: odoo/models.py [hunk 2]
```diff
@@ -227,6 +227,7 @@ class BaseModel(object):
 
     _name = None                # the model name
     _description = None         # the model's informal name
+    _is_business = False        # should be True for important business models
```
Comment (rco-odoo) [inline odoo/models.py]: I prefer
```python
    _main = False               # set it to True on main or principal models
```

File: odoo/addons/base/ir/ir_model.py [hunk 3]
```diff
@@ -54,10 +54,11 @@ class Unknown(models.AbstractModel):
 class IrModel(models.Model):
     _name = 'ir.model'
     _description = "Models"
-    _order = 'model'
+    _order = 'is_business DESC, model ASC'
 
     name = fields.Char(string='Model Description', translate=True, required=True)
     model = fields.Char(default='x_', required=True, index=True)
+    is_business = fields.Boolean('Business Model', help='Used to differentiate important business models from technical and secondary models')
```
Comment (rco-odoo) [inline odoo/addons/base/ir/ir_model.py]: What about adding this field in form views?

File: odoo/models.py [hunk 4]
```diff
@@ -277,17 +277,18 @@ def _field_create(self):
...
         cr.execute(""" UPDATE ir_model
-                       SET name=%(name)s, info=%(info)s, transient=%(transient)s
+                       SET name=%(name)s, is_business=%(is_business)s, info=%(info)s, transient=%(transient)s
                        WHERE model=%(model)s
                        RETURNING id """, params)
         if not cr.rowcount:
-            cr.execute(""" INSERT INTO ir_model (model, name, info, state, transient)
-                           VALUES (%(model)s, %(name)s, %(info)s, %(state)s, %(transient)s)
+            cr.execute(""" INSERT INTO ir_model (model, name, is_business, info, state, transient)
+                           VALUES (%(model)s, %(name)s, %(is_business)s, %(info)s, %(state)s, %(transient)s)
```
Comment (rco-odoo) [inline odoo/models.py]: Wait, don't merge too quick! I have another PR (#15070) that refactors this code!

**Comment by rco-odoo** [comment]
I would like @odony 's opinion on this.

File: odoo/models.py (hunk 2 above)
Comment (tde-banana-odoo) [inline odoo/models.py]: main_business maybe. Main does not mean anything; ir.model is a main model. business is the key word.

File: odoo/models.py (hunk 1 above)
Comment (tde-banana-odoo) [inline odoo/models.py]: You are being a bit psycho-rigid as it is in a separate commit.

File: odoo/models.py (hunk 4 above)
Comment (tde-banana-odoo) [inline odoo/models.py]: First to merge wins :waning_gibbous_moon:

File: odoo/addons/base/ir/ir_model.py (hunk 3 above)
Comment (tde-banana-odoo) [inline odoo/addons/base/ir/ir_model.py]: Seems logical.

File: odoo/models.py (hunk 2 above)
Comment (rco-odoo) [inline odoo/models.py]: What about "main model in application"?
Odoo is an application framework at its core.
"Business" is very ERP-oriented, "Application" is more general purpose.
The model `ir.model` has no direct logic for a regular user: it is fundamentally a technical artifact.

Comment (rco-odoo) [inline odoo/models.py]: Note that `sale.order.line` **is** a business model, but it is a secondary one.
The main or principal model in module `sale` is `sale.order`.
What about "principal" models?

**Comment by odony** [comment]
To use @bst-odoo's wording, I'm not *found of* the terminology here ;-)
Seriously, neither `main` nor `business` sound convincing. That flag also looks out of place in the technical declaration of a model, for some reason. Last but not least, adding an extra flag and an extra field on `ir.model` for such a limited/obscure purpose seems ugly/overkill.

Isn't there some way we could infer the business level of a model based on some existed property?There must be something that sets those "business documents" apart for the technical ones naturally.
How about the fact that they all inherit from `mail.thread`, because users often discuss on them, for example?

Couldn't we instead have a stored computed field on `ir_model` that identifies models with a chatter, and filter/prioritize those however we like depending on the context? (maybe studio already has a field like that BTW, haven't checked)
Another advantage is that custom models with a chatter would automatically be included in "business documents" without users having to manually turn a "magical" flag on.

**Comment by tde-banana-odoo** [comment]
It seems feasible. Probably it won't match 100% of use cases but as this is only used for ordering and showing those models first, it is already a good step forward.

**Comment by rco-odoo** [comment]
@odony @tde-banana-odoo we already have such a field on `ir.model`.
See `web_studio/models/ir_model.py` in `enterprise`.

It was added in Odoo Studio to enable users to create custom models that inherit `mail.thread`.
The field also reflects that property on non-custom models.

We could move the field definition in module `mail`, and override the `_order` on `ir.model` in this extension.

**Comment by rco-odoo** [comment]
See an alternative implementation at #15107.
